File: src/com/l2kt/gameserver/network/serverpackets/FlyToLocation.java
package com.l2kt.gameserver.network.serverpackets;

import com.l2kt.gameserver.model.WorldObject;
import com.l2kt.gameserver.model.actor.Creature;

/**
 * @author KenM
 */
public final class FlyToLocation extends L2GameServerPacket
{
	private final int _destX, _destY, _destZ;
	private final int _chaObjId, _chaX, _chaY, _chaZ;
	private final FlyType _type;
	
	public enum FlyType
	{
		THROW_UP,
		THROW_HORIZONTAL,
		DUMMY; // no effect
	}
	
	public FlyToLocation(Creature cha, int destX, int destY, int destZ, FlyType type)
	{
		_chaObjId = cha.getObjectId();
		_chaX = cha.getX();
		_chaY = cha.getY();
		_chaZ = cha.getZ();
		_destX = destX;
		_destY = destY;
		_destZ = destZ;
		_type = type;
	}
	
	public FlyToLocation(Creature cha, WorldObject dest, FlyType type)
	{
		this(cha, dest.getX(), dest.getY(), dest.getZ(), type);
	}
	
	@Override
	protected void writeImpl()
	{
		writeC(0xC5);
		writeD(_chaObjId);
		writeD(_destX);
		writeD(_destY);
		writeD(_destZ);
		writeD(_chaX);
		writeD(_chaY);
		writeD(_chaZ);
		writeD(_type.ordinal());
	}
}

File: src/com/l2kt/gameserver/network/serverpackets/FlyToLocation.kt
package com.l2kt.gameserver.network.serverpackets

import com.l2kt.gameserver.model.WorldObject
import com.l2kt.gameserver.model.actor.Creature

/**
 * @author KenM
 */
class FlyToLocation(
    cha: Creature,
    private val _destX: Int,
    private val _destY: Int,
    private val _destZ: Int,
    private val _type: FlyType
) : L2GameServerPacket() {
    private val _chaObjId: Int = cha.objectId
    private val _chaX: Int = cha.x
    private val _chaY: Int = cha.y
    private val _chaZ: Int = cha.z

    enum class FlyType {
        THROW_UP,
        THROW_HORIZONTAL,
        DUMMY
    }

    constructor(cha: Creature, dest: WorldObject, type: FlyType) : this(cha, dest.x, dest.y, dest.z, type) {}

    override fun writeImpl() {
        writeC(0xC5)
        writeD(_chaObjId)
        writeD(_destX)
        writeD(_destY)
        writeD(_destZ)
        writeD(_chaX)
        writeD(_chaY)
        writeD(_chaZ)
        writeD(_type.ordinal)
    }
}